### DISCUSSION TITLE
question about the workshop

### DISCUSSION THREAD

**Discussion by Oops-Qiao** [idaholab/moose#22331] [Q&A General] [answered]

**Comment by Oops-Qiao** [comment, marked as answer]
Dear all,

Just a quick question from workshop, in your presentation it is mentioned ADKernel should be used when the PDE is multiplied by both the test function and the gradient of the test function. However, in the following the PDE is only multiplied with gradient of test function, why is the ADKernel is used not the ADKernelGrad one? Thank you very much.
![image](https://user-images.githubusercontent.com/70176014/194641492-33ece76b-2be9-4204-9f38-8d2be06b0cda.png)
![image](https://user-images.githubusercontent.com/70176014/194641564-db118973-9728-43b9-b6ad-71c67d716ff6.png)
![image](https://user-images.githubusercontent.com/70176014/194641578-6652477a-c227-4c94-8c93-c62729a6dc59.png)

**Reply by GiudGiud** [reply]
ADKernel lets you overwrite computeQpResidual
while ADKernelGrad lets you overwrite preComputeQpResidual

the difference is that the latter has a multiplication by the gradient of a test function built in
